google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryImpl.java

```diff
@@ -430,16 +433,39 @@ public InsertAllResponse insertAll(InsertAllRequest request) {
     requestPb.setIgnoreUnknownValues(request.ignoreUnknownValues());
     requestPb.setSkipInvalidRows(request.skipInvalidRows());
     requestPb.setTemplateSuffix(request.getTemplateSuffix());
+    // Using an array of size 1 here to have a mutable boolean variable, which can be modified in
+    // an anonymous inner class.
+    final boolean[] allInsertIdsSet = {true};
     List<Rows> rowsPb = Lists.transform(request.getRows(), new Function<RowToInsert, Rows>() {
       @Override
       public Rows apply(RowToInsert rowToInsert) {
+        allInsertIdsSet[0] &= rowToInsert.getId() != null;
         return new Rows().setInsertId(rowToInsert.getId()).setJson(rowToInsert.getContent());
       }
     });
     requestPb.setRows(rowsPb);
-    return InsertAllResponse.fromPb(
-        bigQueryRpc.insertAll(tableId.getProject(), tableId.getDataset(), tableId.getTable(),
-            requestPb));
+
+    TableDataInsertAllResponse responsePb;
+    if (allInsertIdsSet[0]) {
+      // allowing retries only if all row insertIds are set (used for deduplication)
+      try {
+        responsePb = runWithRetries(
```

Comment:
Can we test the retry?

Reply:
It is very hard to test it in integration tests. In unit tests this behavior is not tested for all other retrieable bigquery methods. The `setUp()` method in `BigQueryImplTest` has this call:
`setRetrySettings(ServiceOptions.getNoRetrySettings())`